### PR TITLE
Add google document AI (second phase)

### PR DESCRIPTION
**Pull request by Andrewed95**
#12 
#13

### CODE REVIEW

File: purchase_invoice.py [hunk 1]
```diff
@@ -875,7 +884,188 @@ def parse_invoice(self,infile,account=None,paid_by_submitter=False,
...
+            [el.get('value') for el in invoice_json['entities'] if el.get('type') == 'total_tax_amount'][0])
+        self.totals[self.default_vat] = float(
+            [el.get('value') for el in invoice_json['entities'] if el.get('type') == 'net_amount'][0])
+
+        if not is_test:
+            if self.check_if_present():
+                return self
+        else:
+            return self
+
```
Comment (tillmo) [inline purchase_invoice.py:1035]: The following lines are duplicated code. Please factor this out.

**Review by tillmo — CHANGES_REQUESTED** [review]

File: purchase_invoice.py [hunk 2]
```diff
@@ -875,7 +884,188 @@ def parse_invoice(self,infile,account=None,paid_by_submitter=False,
...
+        self.assign_default_e_items({self.default_vat: account})
+        return self
+
+    def parse_invoice_json(self, invoice_json, account=None, paid_by_submitter=False, given_supplier=None, is_test=False):
+        self.extract_items = self.update_stock
+        supps = dict(PurchaseInvoice.suppliers)
+        if self.company.name != 'Laden':
+            del supps['Rechnung']
+        try:
+            self.supplier = [el.get('value') for el in invoice_json['entities'] if el.get('type') == 'supplier_name'][0]
```
Comment (tillmo) [inline purchase_invoice.py]: for PreR00492, I get `TypeError: string indices must be integers` here

**Review by tillmo — CHANGES_REQUESTED** [review]

File: purchase_invoice.py [hunk 3]
```diff
@@ -875,7 +884,188 @@ def parse_invoice(self,infile,account=None,paid_by_submitter=False,
...
+            for supplier, info in supps.items():
+                if supplier in self.supplier or self.supplier in supplier:
+                    self.parser = supplier
+                    if info['raw']:
+                        self.raw = True
+                    print("Verwende Rechnungsparser für ", self.supplier)
+                    if given_supplier and self.supplier != given_supplier:
+                        print("abweichend von PreRechnung: ", given_supplier)
+                    self.multi = info['multi']
```
Comment (tillmo) [inline purchase_invoice.py]: you do not need all this stuff. This was for selecting a supplier-specific hand-written parser.

**Review by tillmo — CHANGES_REQUESTED** [review]

File: purchase_invoice.py (hunk 2 above)
Comment (Andrewed95) [reply purchase_invoice.py]: Which command was used? I need it to reproduce this error

File: purchase_invoice.py [hunk 4]
```diff
@@ -875,7 +884,188 @@ def parse_invoice(self,infile,account=None,paid_by_submitter=False,
...
+            [el.get('value') for el in invoice_json['entities'] if el.get('type') == 'total_tax_amount'][0])
+        self.compute_total()
+        return self
+
+    def parse_generic_json(self, invoice_json, default_account=None, paid_by_submitter=False, is_test=False):
+        amount = ""
+        self.parser = "generic"
+        self.extract_items = False
+        self.shipping = 0.0
+        self.date = [el.get('value') for el in invoice_json['entities'] if el.get('type') == 'invoice_date'][0]
```
Comment (tillmo) [inline purchase_invoice.py]: please do not use the first matching element, but the matching element wit highest confidence value (also for the other fields of the invoice object).

**Review by tillmo — CHANGES_REQUESTED** [review]

File: purchase_invoice.py (hunk 2 above)
Comment (tillmo) [reply purchase_invoice.py]: I have done some refactoring. Running play.py shows the error.

Comment (tillmo) [reply purchase_invoice.py]: By the way, if there is no supplier, you can take the one given in the arguments of the method,

Comment (tillmo) [reply purchase_invoice.py]: Is fixed now, see d2d170a083291793b9a5804fa039036ebaa5d4d2

File: purchase_invoice.py (hunk 3 above)
Comment (tillmo) [reply purchase_invoice.py]: I have removed that

File: purchase_invoice.py [hunk 5]
```diff
@@ -875,7 +888,176 @@ def parse_invoice(self,infile,account=None,paid_by_submitter=False,
...
+        if self.update_stock:
+            line_items = [el.get('properties') for el in invoice_json['entities'] if el.get('type') == 'line_item']
+            for line_item in line_items:
+                s_item = SupplierItem(self)
+                for prop in line_item:
+                    if prop['type'] == 'line_item/description':
+                        s_item.description = prop['value']
+                    elif prop['type'] == 'line_item/product_code':
+                        s_item.item_code = prop['value']
+                    elif prop['type'] == 'line_item/quantity' and s_item.qty is None:
```
Comment (tillmo) [inline purchase_invoice.py]: here I get:
AttributeError: 'SupplierItem' object has no attribute 'qty'

File: purchase_invoice.py [hunk 6]
```diff
@@ -875,7 +888,176 @@ def parse_invoice(self,infile,account=None,paid_by_submitter=False,
...
+            account_names = paccs + account_names
+            title = 'Buchungskonto wählen'
+            msg = 'Bitte ein Buchungskonto wählen\n'
+            account = easygui.choicebox(msg, title, account_names)
+            if not account:
+                return None
+        self.assign_default_e_items({self.default_vat: account})
+        return self
+
+    def parse_invoice_json(self, invoice_json, account=None, paid_by_submitter=False, given_supplier=None, is_test=False):
```
Comment (tillmo) [inline purchase_invoice.py]: generally, this method should be restructured. I think parse_google_output and parse_generic_json should be both merged into parse_invoice_json, and duplications removed. Generally, first the code from parse_google_output should be executed. Then, it should be checked whether supplier, date, no, vat and gross_total are all there. If yes, all fine. If not, the form should be shown.

File: purchase_invoice.py (hunk 1 above)
Comment (Andrewed95) [reply purchase_invoice.py:1035]: Fixed

File: purchase_invoice.py (hunk 4 above)
Comment (Andrewed95) [reply purchase_invoice.py]: Done

File: purchase_invoice.py (hunk 5 above)
Comment (Andrewed95) [reply purchase_invoice.py]: Fixed

File: purchase_invoice.py (hunk 6 above)
Comment (Andrewed95) [reply purchase_invoice.py]: Done

**Comment by Andrewed95** [comment]
Dear Till, I sent you a message in freelancer.com. Please check it if possible.
@tillmo

**Pull request merged**